src/hazelcore/extractions/ocaml/OCamlExtraction_Exp.re

```diff
@@ -0,0 +1,270 @@
+// (extracted expression, dynamic/inner type) like ExpandResult
+// Htyp.t is like the ground type
+
+//type extract_result = option((ocaml_exp, HTyp.t));
+type t = (ocaml_exp, HTyp.t)
+and ocaml_exp = string;
+
+/* Theorem: If extract(ctx, d) = Some(ce, ty) and ctx |- d : ty then ce :_ocaml Typ.extract(ty) */
+
+//TODO: replace == with HTyp.consistent
+let rec extract = (~ctx: Contexts.t, ~de: DHExp.t): t =>
+  switch (de) {
+  | EmptyHole(_) => failwith("Exp: Empty Hole")
+  | NonEmptyHole(_) => failwith("Exp: Non-Empty Hole")
+  // used for variables that are keywords in Hazel, like "let" and "case"
+  | Keyword(_) => failwith("Exp: Incomplete Program (Keyword)")
+  // an expression with a free variable cannot be extracted to OCaml because free variables cause type errors in OCaml
+  | FreeVar(_, _, _, _) => failwith("Exp: FreeVar is not allowed")
+  | BoundVar(x) =>
+    //directly lookup type in environment
+    let typ = VarCtx.lookup(Contexts.gamma(ctx), x);
+    switch (typ) {
+    | None => failwith("Exp: BoundVar " ++ x ++ " Not Found")
+    // if the hole type, we don't print the type
+    | Some(t) => (x, t)
+    // if (t == Hole) {
+    //   ("(" ++ x ++ ")", t);
+    // } else {
+    //   ("(" ++ x ++ " : " ++ Typ.extract(~t) ++ ")", t);
+    // }
+    };
+  | Let(dp, de1, de2) =>
+    switch (dp) {
+    | Var(x) =>
+      let ocaml_dp = OCamlExtraction_Pat.extract(~dp);
+      let ocaml_de1 = extract(~ctx, ~de=de1);
+      let ctx' = Contexts.extend_gamma(ctx, (x, snd(ocaml_de1)));
+      let ocaml_de2 = extract(~ctx=ctx', ~de=de2);
+      // use the "let .. = ..;;" format
+      // if it's a recursive function, it's format of let ... = fix ..., we change it to "let rec"
+      //, because the recursive function should be declared by "rec" only in ocaml
+      switch (de1) {
+      | FixF(_, _, _) => (
+          "let rec "
+          ++ ocaml_dp
+          ++ " : "
+          ++ OCamlExtraction_Typ.extract(~t=snd(ocaml_de1))
+          ++ " = "
+          ++ fst(ocaml_de1)
+          ++ ";;\n"
```

Comment:
let expressions end with "in", not ";;", in general -- the ";;" syntax will only work at the very top level